### PR TITLE
Simplify Doxygen config/build

### PR DESCRIPTION
**Pull request by drbenmorgan**
This is a slightly more speculative PR as some of it is style and I'm not an expert on Doxygen config. At least, I think it's tricky to make the latter completely generic...

The PR is in two commits (ignore the Merge branch commits, those should be in master already, so I don't know why Github's included them) and covers
1. Moving the Doxygen CMake module to where it's actually used. This makes the purpose and use a bit clearer for users, plus it permits slightly shorter relative paths to be used. The switch to activate Doxygen is also renamed to use the same style as the `BUILD_TESTS` option. I think that is a bit clearer, and puts everything in the same order when working through the curses/GUI. It _could_ be renamed `BUILD_DOCS` rather than `BUILD_DOXYGEN`.
2. Consistent use of CMake expansion variables in the Doxyfile.in, and using `PROJECT_` rather than `CMAKE_` scope. Some stripping of inc/paths has been added, but I'm less sure about this given the use of currently unused/undocumented `DOXYGEN_` expansion variables. If these are to be used, let me know, and I'll update these as needed (A google suggests these originate in the Gaudi build scripts?) 
3. Move the versioning header into the same location as other project headers. This allows it to be easily added to the library target, so it appears in IDEs (tested on Xcode) and provides an example of integrating/installing a generated source. It also makes the header appear in a more logical location in the documentation. This is in this PR as it otherwise broke the documentation build.

### CODE REVIEW

**Comment by hegner** [comment]
Thanks @drbenmorgan ! There are a few problems with this change. 
1. Mostly OK and good. However, I gave the switch to either enable doxygen or not a project specific name - as to make sure that this switch isn't accidentally forwarded if building against this package with another CMake package later having a switch of the same name. I recall having had a problem once with it. 
2. I agree with the change of the variables; and I am think they are correct. A problem is with the fact that I was assuming that you may want to have multiple subpackage directories living alongside. Your current doxygen changes aren't robust there as "package/include" is hard-coded everywhere. And one shouldn't strip the path too much in these cases anyhow. 
3. Moving the version headers into the "package" dir has the same problem. I was assuming multiple "package" dirs. And thus I left everything which should only exist once into the top dir.

**Comment by drbenmorgan** [comment]
@hegner, no problem, expected there to be some issues!

> Mostly OK and good. However, I gave the switch to either enable doxygen or not a project specific name - as to make sure that this switch isn't accidentally forwarded if building against this package with another CMake package later having a switch of the same name. I recall having had a problem once with it.

O.k., makes sense and I can revert that (but perhaps to `HSFTEMPLATE_BUILD_DOC`?)

> I agree with the change of the variables; and I am think they are correct. A problem is with the fact that I was assuming that you may want to have multiple subpackage directories living alongside. Your current doxygen changes aren't robust there as "package/include" is hard-coded everywhere. And one shouldn't strip the path too much in these cases anyhow.

O.k. on the variables. With the sub packages, do you mean users would create the project then at some later date have a structure like:

```
+- srcdir/
   +- CMakeLists.txt
   +- doc/
   +- Apkg/
   |  +- include/
   |  |  +- APkg/
   |  |     +- Apkg.h
   |  +- src/ 
   +- Bpkg/
      +- include/
      |  +- BPkg/
      |     +- Bpkg.h
      +- src/ 
```

Whilst the hardcoding certainly wouldn't work here, I think the stripping is required otherwise the `#include` paths listed in the generated docs won't match the install layout.

As it is, the output doxygen html (using 1.8.11) from the current master branch will give the absolute local source path in the needed `#include` path (Try creating a package from the template and look at the generated file `YOURBUILDDIR/doxygen/html/d3/db9/classYOURPACKAGENAME_1_1_example.html`).

This may be a case where we're trying to be too general or do too much- at least, I see the template as a starting point and exemplar of standard/good practice. It's not going to cover every single use case or layout style.

For now though I can revert the `Doxyfile.in` with just the variable name updates.

> Moving the version headers into the "package" dir has the same problem. I was assuming multiple "package" dirs. And thus I left everything which should only exist once into the top dir.

Right, I was working to the "single package" use case. I can revert this back for now, though as above it does create incorrect include paths in the docs.

To summarise, I'd propose to:
1. Revert the doc build switch to `HSFTEMPLATE_BUILD_DOCS`
2. Retain `PROJECT_` expansion variables in `Doxyfile.in` but remove stripping/hardcoded `package/include` paths.
3. Move versioning header back.

Let me know what you think!

**Comment by hegner** [comment]
@drbenmorgan - I think you proposal sounds fine.

**Comment by drbenmorgan** [comment]
As the changes were a little involved, and to clarify the task, I reverted the previous set of commits, then reapplied changes from the discussion. I've also updated the template's README to remove the gtest reference and describe how to build docs with reference to Doxygen.

**Comment by hegner** [comment]
Thanks! Looks good

So I believe I can cancel my documentation PR then :-)
